### PR TITLE
Analyzer suppression test fix

### PR DESCRIPTION
**Pull request by LakshanF**
Fix for #2579, Roslyn handles `UnconditionalSuppressMessage ` on its own and we are not planning to add any special handling. The PR addresses the following 3 exceptions from the test bed.

- `SuppressWarningsInMembersAndTypesUsingTarget`: This test case uses [Expression.Call ](https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression.call?view=net-6.0#system-linq-expressions-expression-call(system-type-system-string-system-type()-system-linq-expressions-expression()))which has a RUC but is treated as an intrinsic by the trimmer. Ensured that the analyzer handles this method correctly
- `SuppressWarningsInCompilerGeneratedCode`: Analyzer and Linker differ in the warning code generated. Filed #2583 (and #2587 also has relevant information) for this
- `SuppressWarningsViaXml`: As https://github.com/dotnet/linker/issues/2579#issuecomment-1028549170 discussion notes, we plan on not supporting this in the analyzer, at least for the time being. The test does not run in the analyzer.

### CODE REVIEW

**Comment by vitek-karas** [comment]
> SuppressWarningsInMembersAndTypesUsingTarget: Analyzer was generating couple of 2026 warnings that seems reasonable. Suppressed them

If we think that the warning is valid and linker doesn't produce it then it's a bug in the linker. Somebody needs to investigate why the linker doesn't warn in this case.

**Comment by LakshanF** [comment]
The reason the linker doesn't warn seems to be because we have `SkipKeptItemsValidation `on the test and linker's `ResultChecker `skips validation for the extra 2026 warnings needed. I was being a little lazy trying to understand the goal of this test with this combination of warning mix but will take a closer look at the intent of the test

**Comment by vitek-karas** [comment]
`SkipKeptItemsValidation` should have nothing to do with warnings - this is weird.

**Comment by tlakollo** [comment]
This test is failing because it doesn't expect RUC from Expression.Call(). In runtime Expression.Call() is annotated with RUC (https://github.com/dotnet/runtime/blob/895c99cc502913fd7393b95af7cae5b9e6fbdea3/src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs#L439). Linker has intrinsics for handling it and that's why it doesn't warn. What I think the fix should be:
- Add Expression.Call into RequiresUnreferencedCodeAnalyzer ReportSpecialIncompatibleMembersDiagnostic list of methods, next to MakeGenericMethod and MakeGenericType.

**Comment by tlakollo** [comment]
I was looking that currently the analyzer does the same with calls to `System.Type.GetType` (its annotated with RUC in runtime https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Type.CoreCLR.cs and we have an intrinsic that 'knows better' how to handle it) it might be worth it to just check all intrisics and reuse the logic from the shared project in ReportSpecialIncompatibleMembersDiagnostic.

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs [hunk 1]
```diff
@@ -84,15 +85,22 @@ protected override ImmutableArray<ISymbol> GetSpecialIncompatibleMembers (Compil
 				incompatibleMembers.AddRange (methodInfoType.GetMembers ("MakeGenericMethod").OfType<IMethodSymbol> ());
 			}
 
+			var expressionCallType = compilation.GetTypeByMetadataName ("System.Linq.Expressions.Expression");
+			if (expressionCallType != null) {
+				incompatibleMembers.AddRange (expressionCallType.GetMembers ("Call").OfType<IMethodSymbol> ()
+					.Where (m => m.Parameters.Length == 4 && m.Parameters[0] is IParameterSymbol parameterSymbol
```
Comment (tlakollo) [inline src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:91]: The list of incompatible member purposes is to say "RequiresUnreferencedCodeAnalyzer please don't analyze this call" then the dataflow will handle the intrinsic and decide if it needs to warn or not. You are restricting the number of parameters and type of the parameter, which means that RequiresUnreferencedCodeAnalyzer will only bypass that case and for everything else will warn. Im afraid that dataflow will do something similar and we end up generating two warnings, I think what we want is:
```diff
- incompatibleMembers.AddRange (expressionCallType.GetMembers ("Call").OfType<IMethodSymbol> ()
-					.Where (m => m.Parameters.Length == 4 && m.Parameters[0] is IParameterSymbol parameterSymbol
-					&& parameterSymbol.Type?.ContainingNamespace?.Name == "System" && parameterSymbol.Type?.Name == "Type"));
+ incompatibleMembers.AddRange (expressionCallType.GetMembers ("Call").OfType<IMethodSymbol> ());
```

Comment (LakshanF) [reply src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:91]: unfortunately, there are lots of these Call methods and only 1 should be handled intrinsically. The code mimics the [linker ](https://github.com/dotnet/linker/blob/feature/damAnalyzer/src/ILLink.Shared/TrimAnalysis/Intrinsics.cs#L59-L63)version. There are some other expression members that need special handling and I'm taking a look at those as part of #2578

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs [hunk 2]
```diff
@@ -84,15 +85,22 @@ protected override ImmutableArray<ISymbol> GetSpecialIncompatibleMembers (Compil
...
 			return incompatibleMembers.ToImmutable ();
 		}
 
 		protected override bool ReportSpecialIncompatibleMembersDiagnostic (OperationAnalysisContext operationContext, ImmutableArray<ISymbol> specialIncompatibleMembers, ISymbol member)
 		{
 			if (member is IMethodSymbol method && ImmutableArrayOperations.Contains (specialIncompatibleMembers, member, SymbolEqualityComparer.Default) &&
-				(method.Name == "MakeGenericMethod" || method.Name == "MakeGenericType")) {
-				// These two RUC-annotated APIs are intrinsically handled by the trimmer, which will not produce any
-				// RUC warning related to them. For unrecognized reflection patterns realted to generic type/method
+				(method.Name == "MakeGenericMethod" || method.Name == "MakeGenericType" || method.Name == "Call")) {
```
Comment (sbomer) [inline src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:101]: Why do we need to check the method name if we are already comparing with `specialIncompatibleMembers`? I think we could get rid of the name check here. cc @tlakollo

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs (hunk 1 above)
Comment (sbomer) [reply src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:91]: I think https://github.com/dotnet/linker/issues/2578 will require a more general solution but I am fine with this change doing the minimum necessary to get the suppression tests working.

**Review by sbomer — COMMENTED** [review]
Thanks @LakshanF!

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs (hunk 2 above)
Comment (LakshanF) [reply src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs]: missed this duplicate check, thanks!

**Pull request merged**